java-frontend/src/main/java/org/sonar/java/se/checks/CustomUnclosedResourcesCheck.java

```diff
@@ -0,0 +1,453 @@
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.se.checks;
+
+import com.google.common.collect.ImmutableList;
+import org.sonar.check.Priority;
+import org.sonar.check.Rule;
+import org.sonar.check.RuleProperty;
+import org.sonar.java.se.CheckerContext;
+import org.sonar.java.se.ProgramState;
+import org.sonar.java.se.constraint.ConstraintManager;
+import org.sonar.java.se.constraint.ObjectConstraint;
+import org.sonar.java.se.symbolicvalues.SymbolicValue;
+import org.sonar.plugins.java.api.semantic.Symbol;
+import org.sonar.plugins.java.api.semantic.Type;
+import org.sonar.plugins.java.api.tree.Arguments;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.MemberSelectExpressionTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.NewClassTree;
+import org.sonar.plugins.java.api.tree.ReturnStatementTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.squidbridge.annotations.NoSqale;
+import org.sonar.squidbridge.annotations.RuleTemplate;
+
+import javax.annotation.Nullable;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+@Rule(
+  key = "S3546",
+  name = "Resources as defined by user should be closed",
+  priority = Priority.BLOCKER,
+  tags = {"bug", "cert", "cwe", "denial-of-service", "leak", "security"})
+@RuleTemplate
+@NoSqale
+public class CustomUnclosedResourcesCheck extends SECheck {
+
+  private static final String ARGUMENT_DESCRIPTION = " If the signature is important, " +
+    "then it must be supplied by listing the full qualified names of each argument, coma delimited within parentheses.";
+
+  private enum Status {
+    OPENED, CLOSED
+  }
+
+  @RuleProperty(
+    key = "constructor",
+    description = "List of classes (full qualified name) used to create instances that need to be closed." + ARGUMENT_DESCRIPTION)
+  public String constructor = "";
+
+  @RuleProperty(
+    key = "factoryMethod",
+    description = "List of classes (full qualified name#methodName) used to open instances that need to be closed." + ARGUMENT_DESCRIPTION)
```

Comment:
update description according to RSPEC https://jira.sonarsource.com/browse/RSPEC-3546